File: services/server/src/auth/controller/auth.controller.ts
import {
  Controller,
  Post,
  Get,
  Patch,
  Body,
  UseGuards,
  HttpCode,
  HttpStatus,
  NotImplementedException,
} from "@nestjs/common";
import { ApiTags, ApiOperation, ApiBearerAuth } from "@nestjs/swagger";
import { AuthService } from "@/auth/service/auth.service";
import { UpdateUserData, UsersService } from "@/users/service/users.service";
import {
  RegisterRequest,
  LoginRequest,
  RefreshTokenRequest,
  UpdateProfileRequest,
  ChangePasswordRequest,
  ForgotPasswordRequest,
  ResetPasswordRequest,
} from "@/auth/controller/auth-request";
import { JwtAuthGuard } from "./jwt-auth.guard";
import { CurrentUser } from "./current-user.decorator";
import { Public } from "./public.decorator";

@ApiTags("auth")
@Controller("auth")
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly usersService: UsersService,
  ) {}

  @Public()
  @Post("register")
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: "Register" })
  async register(@Body() registerDto: RegisterRequest) {
    const result = await this.authService.register(registerDto);
    return {
      user: {
        id: result.user.id,
        email: result.user.email,
        username: result.user.username,
        phone: result.user.phone,
        avatar: result.user.avatar,
        status: result.user.status,
      },
      token: result.tokens.accessToken,
      refreshToken: result.tokens.refreshToken,
    };
  }

  @Public()
  @Post("login")
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Login" })
  async login(@Body() loginDto: LoginRequest) {
    const result = await this.authService.login(loginDto);
    return {
      user: {
        id: result.user.id,
        email: result.user.email,
        username: result.user.username,
        phone: result.user.phone,
        avatar: result.user.avatar,
        status: result.user.status,
      },
      token: result.tokens.accessToken,
      refreshToken: result.tokens.refreshToken,
    };
  }

  @Public()
  @Post("refreshToken")
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Refresh tokens" })
  async refreshToken(@Body() refreshTokenDto: RefreshTokenRequest) {
    const tokens = await this.authService.refreshToken(
      refreshTokenDto.refreshToken,
    );
    return {
      token: tokens.accessToken,
      refreshToken: tokens.refreshToken,
    };
  }

  /** @deprecated Prefer POST /auth/refreshToken */
  @Public()
  @Post("refresh-token")
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: "Refresh tokens (legacy path)" })
  async refreshTokenLegacy(@Body() refreshTokenDto: RefreshTokenRequest) {
    return this.refreshToken(refreshTokenDto);
  }

  @UseGuards(JwtAuthGuard)
  @Post("logout")
  @HttpCode(HttpStatus.NO_CONTENT)
  @ApiBearerAuth()
  @ApiOperation({ summary: "Logout" })
  async logout() {
    return;
  }

  @UseGuards(JwtAuthGuard)
  @Get("me")
  @ApiBearerAuth()
  @ApiOperation({ summary: "Get current user" })
  async getCurrentUser(@CurrentUser() user: unknown) {
    return { user };
  }

  @UseGuards(JwtAuthGuard)
  @Patch("profile")
  @ApiBearerAuth()
  @ApiOperation({ summary: "Update profile" })
  async updateProfile(
    @CurrentUser() user: { id: string },
    @Body() updateProfileDto: UpdateProfileRequest,
  ) {
    const data: UpdateUserData = {};
    if (updateProfileDto.username !== undefined)
      data.username = updateProfileDto.username;
    if (updateProfileDto.phone !== undefined)
      data.phone = updateProfileDto.phone;
    if (updateProfileDto.status !== undefined)
      data.status = updateProfileDto.status;
    if (updateProfileDto.avatar !== undefined)
      data.avatar = updateProfileDto.avatar;

    const updatedUser = await this.usersService.updateUser(user.id, data);
    return { user: updatedUser };
  }

  @UseGuards(JwtAuthGuard)
  @Post("changePassword")
  @ApiBearerAuth()
  @ApiOperation({ summary: "Change password" })
  async changePassword(
    @CurrentUser() _user: unknown,
    @Body() _changePasswordDto: ChangePasswordRequest,
  ) {
    throw new NotImplementedException("Change password is not implemented");
  }

  @Public()
  @Post("forgotPassword")
  @ApiOperation({ summary: "Forgot password" })
  async forgotPassword(@Body() _forgotPasswordDto: ForgotPasswordRequest) {
    throw new NotImplementedException("Forgot password is not implemented");
  }

  @Public()
  @Post("resetPassword")
  @ApiOperation({ summary: "Reset password" })
  async resetPassword(@Body() _resetPasswordDto: ResetPasswordRequest) {
    throw new NotImplementedException("Reset password is not implemented");
  }
}

File: src/main/java/com/chat/auth/controller/AuthController.java
package com.chat.auth.controller;

import com.chat.auth.controller.AuthRequests.ChangePasswordRequest;
import com.chat.auth.controller.AuthRequests.ForgotPasswordRequest;
import com.chat.auth.controller.AuthRequests.LoginRequest;
import com.chat.auth.controller.AuthRequests.RefreshTokenRequest;
import com.chat.auth.controller.AuthRequests.RegisterRequest;
import com.chat.auth.controller.AuthRequests.ResetPasswordRequest;
import com.chat.auth.controller.AuthRequests.UpdateProfileRequest;
import com.chat.auth.service.AuthService;
import com.chat.users.controller.UserResponse;
import com.chat.users.service.UsersService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/v1/auth")
public class AuthController {

  private final AuthService authService;
  private final UsersService usersService;

  public AuthController(AuthService authService, UsersService usersService) {
    this.authService = authService;
    this.usersService = usersService;
  }

  @PostMapping("register")
  @ResponseStatus(HttpStatus.CREATED)
  public Map<String, Object> register(@Valid @RequestBody RegisterRequest body) {
    return authService.register(body.email(), body.password(), body.username(), body.phone());
  }

  @PostMapping("login")
  public Map<String, Object> login(@Valid @RequestBody LoginRequest body) {
    return authService.login(body.email(), body.password());
  }

  @PostMapping("refreshToken")
  public Map<String, String> refresh(@Valid @RequestBody RefreshTokenRequest body) {
    return authService.refresh(body.refreshToken());
  }

  @PostMapping("logout")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void logout(HttpServletRequest request) {
    authService.logout(request.getHeader("Authorization"));
  }

  @PostMapping("changePassword")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void changePassword(
      Authentication authentication, @Valid @RequestBody ChangePasswordRequest body) {
    authService.changePassword(
        authentication.getName(), body.currentPassword(), body.newPassword());
  }

  @PostMapping("forgotPassword")
  public Map<String, Object> forgotPassword(@Valid @RequestBody ForgotPasswordRequest body) {
    return authService.forgotPassword(body.email());
  }

  @PostMapping("resetPassword")
  @ResponseStatus(HttpStatus.NO_CONTENT)
  public void resetPassword(@Valid @RequestBody ResetPasswordRequest body) {
    authService.resetPassword(body.token(), body.newPassword());
  }

  @GetMapping("me")
  public Map<String, UserResponse> me(Authentication authentication) {
    return Map.of("user", usersService.getById(authentication.getName()));
  }

  @PatchMapping("profile")
  public Map<String, UserResponse> profile(
      Authentication authentication, @RequestBody UpdateProfileRequest body) {
    UserResponse user =
        usersService.updateProfile(
            authentication.getName(), body.username(), body.phone(), body.status(), body.avatar());
    return Map.of("user", user);
  }
}
